### ISSUE TITLE
Separate `Show Grid` and `Enable Snap` toggles in the Visual Script editor

### ISSUE THREAD

**Issue by jasperbrooks79** [godotengine/godot-proposals#2111]
### Describe the project you are working on
A Tomb Raider game, with epic code and, even epicer beasts and, monsters . . <3 . .

### Describe the problem or limitation you are having in your project
In visual script, it's a bit annoying, when I disable the grid, that snapping also is turned off, and the boxes go sailing away, into the sky or, waters . . :O I was wondering, if in visual script, the ' show grid ' and, ' use snap ' could, not be made into separate icons, or somehow not be, connected, as it would make visual script, even nicer, to use ya ya . .

My new visual script theme, for Godot, personal gold, so cool . .

![Skærmbillede (865)](https://user-images.githubusercontent.com/61069740/104394316-a8067800-5546-11eb-9c61-88dcb53cb768.png)

No stress, just fun . . But, now I need the last feature, to be able to turn off ' grid ', but still have snap, please . .

![Skærmbillede (873)](https://user-images.githubusercontent.com/61069740/104394362-c53b4680-5546-11eb-9f84-573c4859b7b3.png)

<img width="488" alt="Skærmbillede (873)" src="https://user-images.githubusercontent.com/61069740/104394429-e439d880-5546-11eb-8da5-af666a326a80.png">

The LAST thing is, the ' grid ' size ( default 20 ), it would be cool if there was a place, to set that globally, for all visual scripts, right now, even after setting it, when re-opening the same script, it goes back to 20 ( default ), by accident . .

Please, fix these issues or, problems, I can't sleep or, think, until it's done, ya ya . .

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Just the above, soon, or hurry . . More features, more code, or there riots . .

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
Just, the above two ' fixes ', to make the user - experience smooth, as silk, on a pillow, that's all . . . <3 :O . . .

### If this enhancement will not be used often, can it be worked around with a few lines of script?
I can't change the engine, as a new user, not yet . . Please, consider . .

### Is there a reason why this should be core and not an add-on in the asset library?
Well, this is always difficult to answer, I'm not sure . . <3 <3 . . .

**Comment by jasperbrooks79** [comment]
Also, if one could set grid-style to ' dotted ' and, set the opacity, and disable the ' vertical ' lines, so it looks like . .

![2021-01-13 0248](https://user-images.githubusercontent.com/61069740/104395790-c621a780-5549-11eb-9052-d89373d61924.png)

like, graph paper . . <3 . .

basically, so one can enable, or disable ' large lines ', both vertical and, horizontal, and the smaller ' lines ', also, vertical or horizontal . . :O . . . Thanks . .

To get that, custom look <3

**Comment by jasperbrooks79** [comment]
basically, this . . >>
<img width="764" alt="Skærmbillede (874)" src="https://user-images.githubusercontent.com/61069740/104396209-84ddc780-554a-11eb-8720-533b23696ab8.png">

**Comment by jasperbrooks79** [comment]
Also, a ' toggle ' to set text in graph node headers, to italic . . :O :O . .

Over - all, a more ' custom ' way, to set visual script canvas and, grid looks, so one can make it, ' just ' right, I'll find the right settings later, so if one could disable or, enable vertical and, horizontal lines, and set opacity of them, that'd be amazing, maybe like there's an ' editor settings ', where one can set theme, simply, more options, or so . . Then, I can make the most bad - ass, cool visual script experience, in the world <3 . . Thanks . . . .

**Comment by jasperbrooks79** [comment]
and, some of these, for making it really shine, for when grid is on or, not . . >>
![2021-01-13 0316](https://user-images.githubusercontent.com/61069740/104397736-aee4b900-554d-11eb-9eaa-597bcc8296f9.png)

so, we can make some really blazing text headers . . <3

here, or close . . .
![2021-01-13 0316](https://user-images.githubusercontent.com/61069740/104398003-39c5b380-554e-11eb-84cc-30ec43f6186e.png)

not quite, the right font, but it'll look cool, when I'm done, smooth as, butter and, cream, on Sunday <3

Thanks . . .

**Comment by jasperbrooks79** [comment]
anyway, I guess what is really needed, is a new, separate font, for graph node headers, that's all . . 

here, is looking really nice, but if graph node headers could be italic, that'd be sweet, no more . .

<img width="633" alt="Skærmbillede (878)" src="https://user-images.githubusercontent.com/61069740/104399081-43501b00-5550-11eb-94a4-093ff6122beb.png">

Thanks <3 . . .

**Comment by jasperbrooks79** [comment]
Add, also a way to ' color ' comment boxes, so they look like <3 . .
<img width="1447" alt="Skærmbillede (879)" src="https://user-images.githubusercontent.com/61069740/104402839-7944cd80-5557-11eb-9ae0-8e224e7e43bc.png">

post it <3 . .

![2021-01-13 0427](https://user-images.githubusercontent.com/61069740/104402894-9da0aa00-5557-11eb-9711-f76c2f355c89.jpg)

And, a way to just ' write ' on the canvas, with any font . .

floating text . . . >>

![2021-01-13 0427](https://user-images.githubusercontent.com/61069740/104403122-1bfd4c00-5558-11eb-9776-b40b851cf7f1.jpg)

the fonts aren't the best, but one could write text, anywhere, and pick font and, color of text, for when navigating, a script <3 Thanks . . Trust me, it'll be so good, you won't believe it . . . <3 <3 . .

**Comment by jasperbrooks79** [comment]
@Calinou sorry, to disturb, I added some stuff to it, I know not to disturb engine devs, your time is precious, I hope we can get this, in Godot, to make it ' rock ', so hard . .

<3 . .

**Comment by jasperbrooks79** [comment]
okay, a way to make the comments look, like this . . so, one can also change color, of the HEADER, of the comment . .
<img width="1216" alt="Skærmbillede (880)" src="https://user-images.githubusercontent.com/61069740/104404151-4ea84400-555a-11eb-9784-da8d04c7652d.png">

the last really cool thing, was if one could divide the canvas into like, more viewports, and then drag or, connect nodes, between them, for faster coding, or so . . so, one could connect two nodes, that were far apart or, so . . and, select nodes and, move them into other viewports, so to move a cluster of nodes, one can select them in one viewport, and drag them into another part, of the script . . and, the last thing, is a window, where one can also write code, at same time, maybe . . to make, editing easier, faster, or so <3 . . .

**Comment by jasperbrooks79** [comment]
if we could get these things, so I could experiment with them, I could make a kick - ass visual script theme, and make many, many users more happy, in using it, etc . . but, it needs to be so, I can set the colors and, fonts, also sizes, very custom . . :O :O . . . Thanks . .

gonna make it shine, like a mother, ya ya . . .

**Comment by jasperbrooks79** [comment]
the last thing, add so one can have ' layers ', in the script . .

<img width="1167" alt="Skærmbillede (881)" src="https://user-images.githubusercontent.com/61069740/104404949-228dc280-555c-11eb-8276-030629cb1dee.png">

so, the layers beneath gets faded or, 'dithered', or so . .

![2021-01-13 0500](https://user-images.githubusercontent.com/61069740/104405017-4c46e980-555c-11eb-95e2-cd31db74802d.png)

having these things, would take visual script, to another level, given some time, I'd make the ultimate version, possible . . like, a breeze, ya ya . .

**Comment by jasperbrooks79** [comment]
the effects and, so aren't 100 % there, I'm not so good at drawing or, finding the right images, sry . . . <3 . . .

**Comment by jasperbrooks79** [comment]
the last thing would be, to be able to go into a semi or, so 3D view, of the scripts . .

<img width="524" alt="2021-01-13 0515" src="https://user-images.githubusercontent.com/61069740/104405936-6da8d500-555e-11eb-94e5-29650ca7816f.png">

so, you can drag and, drop or, connect nodes, between layers, may-be, and where the ' box ' 3d effects, fades, and you can rotate the canvas a bit, to get a ' good ' view . . so, you can drop nodes between layers like, bricks, or simply the ability to, create nodes in each layer, or have a button to, ' move 1 layer down ' or, up, or so . .

that'd be, an amazing, nice Visual script experience, really optimized, and given some time, I could make it as rock - solid, as written code, since this answers all the limitations, of visual script, compared to written code, ALMOST, not entirely, Thanks . . .

**Comment by jasperbrooks79** [comment]
also, having the ability to, drag out an arrow, and make it continue, in another place, would be nice . . <3

<img width="524" alt="2021-01-13 0523" src="https://user-images.githubusercontent.com/61069740/104406405-81a10680-555f-11eb-9d3e-95ae22cf816a.png">

this should, make visual script, as powerful as, written code, and all the other engines, would lose their collective shit, in admiration <3 . . Thanks, also I'd like it, for my own stuff, if possible . .

**Comment by jasperbrooks79** [comment]
last, when you select something, in the ' 3d view ' effect, it looks, like this, or so . . .

<img width="448" alt="2021-01-13 0542" src="https://user-images.githubusercontent.com/61069740/104407584-33413700-5562-11eb-86ce-d0d3d1eeb6dc.png">

like, a halo or, so, since it's in Godot <3 ya ya . .

that'd be, cool, and people could combine the things, to make their own, special, personal theme or, version, maybe not even using the 3D effect, or pick different dithering effects, for the layers . .

<3

**Comment by jasperbrooks79** [comment]
ps. as reliable, as a saint or, angel, for God-ot <3 Thanks . .

I'm an autist, I do stuff like this, for some reason, Thanks . . for, fun or, such, my 'hobby' . .

Thanks <3 <3

Smoke . . :OO <3 . . .

**Comment by jasperbrooks79** [comment]
last, arrows would have to be turned into 3D splines, instead of 2D splines, then, I think it works, from all angles or, so . .

Thanks . .

( sorry, I'm an autist, have hard OCD, I come up with really weird stuff, or so, hope you, like it, it will be really cool, no matter how you use it, as well . . . <3

**Comment by jasperbrooks79** [comment]
for it to work, you'd have to make some 2D nodes, then make some under 3D nodes, that ' follow ' the 2D stuff, and then just ' pan ' the 2D layer, for it to be ' pixel - perfect ', not sure . .

it'd look good, in ' orthographic ' view or, so . .

something, like this, but with ' locked ' 2D effect >

![2021-01-13 0620](https://user-images.githubusercontent.com/61069740/104409883-84075e80-5567-11eb-82e4-363a67e440fa.gif)

combining 2D rendering, with 3D, in some ' weird ' way, effect might also be cool, in later 2D games or, so cool . .

**Comment by jasperbrooks79** [comment]
I think, you have to combine a 2D + 3D parallax effect, but since Godot can already render 3D, in the editor, it should be able to run well, on any modern / current computer, with a 3D card, and if it's too slow, one can just not use it, as well . . .

**Comment by jasperbrooks79** [comment]
last, to make connecting, between layers easier, if one could select on output port and, one input port and, click connect, so it ' just ' works or, so . .

![2021-01-13 0659](https://user-images.githubusercontent.com/61069740/104412599-e0b94800-556c-11eb-8163-dcccea151148.jpg)

however, since most connections will be made, in the SAME layers, it will only be needed, for special cases or, such . .

Thanks . . .

**Comment by jasperbrooks79** [comment]
last, if like in Bolt, the ' script ' was live converted to, C# or, GDNative, visual script would be amazing, not sure if one can convert to GDNative, from visual script . . but, Bolt sort of, has a method of, converting the visual script into C#, on the fly, that'd be amazing, and make visual script into an un-heard of beast, possibly, the best in the world or, close . . <3

**Comment by jasperbrooks79** [comment]
if you want to make the effect, for a ParallaxPerspective3D node, you have to place a camera, for each background layer, and render the objects in each layer, and then ' paint ' the image, on the background, like a Viewport texture, might be cool in games, the disadvantage is you have to make each layer separately, and then only have each camera render those layers . .

<img width="601" alt="Skærmbillede (884)" src="https://user-images.githubusercontent.com/61069740/104414210-29becb80-5570-11eb-9739-d441288efd02.png">

anyway, that proves the effect is possible, for visual script or, so . .

<3 . .

**Comment by jasperbrooks79** [comment]
last part is, the ' 2D face ', of boxes in background layers, will have to be turned into textures and, then put on background boxes, since the camera won't pan very far, to each side, it should be efficient . .

when it's done, if possible, it'll blow all other visual script solutions out of the water, like a breeze or, so . .

Thanks :OO . . .

**Comment by jasperbrooks79** [comment]
here's size, of the textures, that need to be generated . . >

<img width="604" alt="Skærmbillede (885)" src="https://user-images.githubusercontent.com/61069740/104416070-693ae700-5573-11eb-9038-631e1d19c55b.png">

I THINK, it would run fast <3

**Comment by jasperbrooks79** [comment]
to reduce lag, one could only render maybe, 200 pixels to each side, of the screen or, compress the textures, some - what, as the ' dithering ' effect will mean, super - clear details, won't be seen . .

also, the textures could be generated, each time a new box is made, and then store them, on the SSD, or so, for a faster experience, or, they could be generated each time a script is opened, but this would take time, for larger scripts . . the ' textures ' could then be deleted, before exporting and, stored in a single file, like a .zip, or so . .

or, they could be generated when opening each script, if that is fast, or so . .

maybe, the GPU can do some fancy stuff, to render them fast, in parellel like, GPU particles <3

**Comment by jasperbrooks79** [comment]
also, if this works ( it will ), then every-one will want to start with using Godot, as their first game engine, as it's the best experience, around . . might be a nice boost, over Patreon so, you could hire more people, partial or, - full time, every-body wins ya ya . . 

<3 . . .

**Comment by jasperbrooks79** [comment]
you could make a little progress bar, at the bottom or, such . . that shows, when the entire script, is rendered in memory, I assume it will happen fast, but just to be sure, it's all stable, reliable . . . <3

**Comment by jasperbrooks79** [comment]
anyway, having a way to see what the visual code looks, like in GDScript and, may-be C#, would be enough . .

then, it'd be cool way, to get into code, as well . . <3 <3

**Comment by jasperbrooks79** [comment]
the last thing is, since there will be ' layers ', there needs to be a way, to move an entire layer, the ' selected ' one, like the ' entire ' screen, it'd be really nice, thanks . . <3

**Comment by bluenote10** [comment]
@jasperbrooks79 I just wanted to make you aware that people watching the repository get an email + GitHub notification every time you add a new message. You are by far dominating my GitHub email notifications these days, and I'm watching many big repositories. On this issue alone ~29 messages in a row. In fact, it is not unlikely that I have never received more emails from anyone else than from you :wink:

Perhaps you could make your proposals more community friendly by
- Using the edit button instead of posting new messages.
- Before posting a proposal, think about a good and concise way to present it.
- Leave out linguistic fillers. Your proposals are often very hard to follow, because the point you are making is buried in a lot noise.
- Stick to standard typography. I often have a hard time e.g. understanding your usage of `'` (if you put spaces on both sides it is so hard to see to which side the quote goes). Similarly if you put spaces around the dots all the time, there can be really weird line breaks. The standard way to type ellipsis `...` even saves you some key strokes :wink:. The ubiquitous "less than 3" statements can also be confusing, for instance when you say something like `... Godot 4 <3`. For a moment I thought you mean Godot 4 is worse than Godot 3.

Any attempt to make your proposals more accessible would be highly appreciated :heart:. The energy you put into this is inspiring, and I'm sure there is a lot of value in your proposals, so it would be a pity if they don't get the attention they deserve!

**Comment by Zireael07** [comment]
<3 is just a way to write the heart emoji when someone doesn't have the emoji font.

But otherwise very good points, especially the one about editing instead of posting new messages.

**Comment by jasperbrooks79** [comment]
I didn't think, you get e-mail notifications, it's not me sending them :(

Sorry, it can be confusing, also, the excitement, and so much . . .

**Comment by jasperbrooks79** [comment]
if you did this, Godot's visual script would leave all other engines in the dust, including Unreal . . I'm very close to make a really good, stable visual script theme, looks like this, so far . .

![Skærmbillede (895)](https://user-images.githubusercontent.com/61069740/104675001-e6d23480-56e4-11eb-8a51-f032e6cae106.png)

and . .
![Skærmbillede (900)](https://user-images.githubusercontent.com/61069740/104675101-197c2d00-56e5-11eb-8c8d-7c08e9b6f1a1.png)

if I get the option, to customize all these things, it will run like butter on, a grill / BBQ <3

![Skærmbillede (901)](https://user-images.githubusercontent.com/61069740/104675136-2ac53980-56e5-11eb-8d6b-e47950d330ad.png)

if you want the theme, hex for dark color = 121212 . .
dark yellow : c2ac1b . .
<img width="1061" alt="Skærmbillede (904)" src="https://user-images.githubusercontent.com/61069740/104675391-b50d9d80-56e5-11eb-94ed-20ed23f6e855.png">

and, for main lines use same yellow color, on the grid, but for ' smaller ' lines, use this . . >>
<img width="937" alt="Skærmbillede (906)" src="https://user-images.githubusercontent.com/61069740/104675695-467d0f80-56e6-11eb-82ae-20cadc3947af.png">

the dark color = 121212 . . was hoping, it could become a new theme, called ' Visual Scripting ' <3 . . Thanks . . .

**Comment by jasperbrooks79** [comment]
simply made, from this . . <3
![2021-01-15 0411](https://user-images.githubusercontent.com/61069740/104676566-eb4c1c80-56e7-11eb-8f8a-bd4212da5499.jpg)

So, there, hope you like, good fun, ya ya . .

<3 . .

**Comment by jasperbrooks79** [comment]
last thing, we need are some pins or, so . .
![2021-01-15 0757](https://user-images.githubusercontent.com/61069740/104691999-80124280-5707-11eb-8d26-834fc5c6661a.jpg)

![2021-01-15 0758](https://user-images.githubusercontent.com/61069740/104692009-83a5c980-5707-11eb-9598-608b76752a09.jpg)

so, we can do this, or so . .
![2021-01-15 0759](https://user-images.githubusercontent.com/61069740/104692105-adf78700-5707-11eb-8a24-0a099d1581dc.jpg)

or, pin notes . . <3
![2021-01-15 0800](https://user-images.githubusercontent.com/61069740/104692152-c5cf0b00-5707-11eb-8271-15cb03a242d1.jpg)

the should be made in 3D, so when one uses the ' 3D ' view, if possible, they look ' 3D ' . .

that's a, nice scripting experience, real cool <3 Thanks . .

**Comment by jasperbrooks79** [comment]
or, so it can look real ' conspiracy ' and, one can connect ' meta ' ideas, on the canvas, for fun <3 <3
![2021-01-15 0802](https://user-images.githubusercontent.com/61069740/104692363-22cac100-5708-11eb-9d05-21f6e38a1427.jpg)

that way, one can ' box ' in ideas, and draw connections, etc . .

**Comment by jasperbrooks79** [comment]
last, maybe ability to put ' photos ' on it, like this . .
![2021-01-15 0820](https://user-images.githubusercontent.com/61069740/104693923-b9987d00-570a-11eb-862b-8c11ac141fbb.jpg)

so, it feels like solving a crime or, ' solving ' a difficult case or, problems <3

**Comment by jasperbrooks79** [comment]
last, if there is a team, with coders and, visual scripters, they'd get asked, by the visual scripters, to use Godot, as it's the best place, to do that, indirectly, it could make Godot more popular, among written coders, so it's win - win, when it's all done and, humming, like a Ferrari or, two . . Thanks . . <3

**Comment by pycbouh** [comment]
Jasper, please keep proposals to a single concise idea. You've started with a simple and reasonable request to divorce snapping and grid visualization from each other and give them dedicated buttons. You've added that it might be useful to only show horizontal or vertical lines, which is, alright, but questionable.

But then you went on a tangent about other features, unrelated features. I understand your excitement, but those features are now lost to the amount of messages you leave us 🙂 It makes it that much harder to evaluate your proposals like that, and some of them are actually very useful, valuable feedback from a person who actually uses visual scripting. Can you please do me a favor and keep your ideas in short, targeted individual issue tickets? And if you have more to add *on the subject* of a proposal, please try to edit your original message to explain more details there.

**Comment by jasperbrooks79** [comment]
Ok, I'm sorry, I learned 1 day ago, each time I post here, someone gets a new e-mail, didn't know you could subscribe to gitHub, didn't think it was like that, sorry . . <3

**Comment by Xrayez** [comment]
> basically, this . . >>
> 
> <img alt="Skærmbillede (874)" width="764" src="https://user-images.githubusercontent.com/61069740/104396313-b22a7580-554a-11eb-9418-d215b8d8f10f.png">
> 
> having these settings, maybe in a ' menu ' or, a few buttons or, drop - down buttons, would make visual script canvas ( maybe also visual shaders ), look more appealing, or fun, totally custom, you decide, what's best <3 . . Thanks . . .

This ability was discussed as part of #2662 proposal (via `_draw_line()` virtual method and whatnot). In theory, all those grid-based classes with snapping could be refactored to use a general-purpose node which could allow customizations like this everywhere.

Regarding the initial proposal as described in the OP, yeah I think it makes sense to have two separate buttons for this.

**Comment by YuriSizov** [comment]
Closing this proposal as we have [officially discontinued the VisualScript module](https://github.com/godotengine/godot/pull/64822). It may return in future as an engine extension, then all the feedback can be reevaluated by maintainers of that extension.

Thanks for your suggestion nonetheless!

**Comment by Zireael07** [comment]
Actually, this proposal also applies to visual shaders (it's generally a limitation/annoyance of the graph itself)

**Comment by YuriSizov** [comment]
Please open a new proposal so it's easier to discuss it in the context of visual shaders.
